langgraph_plugin/human_in_the_loop/approval_graph_interrupt/workflow.py

```
@@ -0,0 +1,241 @@
+        self._app: Any = None  # Store runner for visualization queries
+
+    @workflow.signal
+    def provide_approval(self, response: dict[str, Any]) -> None:
+        """Signal to provide approval response.
+
+        Args:
+            response: Dict with 'approved' (bool), 'reason' (str), 'approver' (str)
+        """
+        self._approval_response = response
+
+    @workflow.query
+    def get_pending_approval(self) -> dict[str, Any] | None:
+        """Query to get the current pending approval request.
+
+        Returns:
+            The interrupt value containing approval request details, or None.
+        """
+        return self._interrupt_value
+
+    @workflow.query
+    def get_status(self) -> str:
+        """Query to get the current workflow status."""
+        if self._interrupt_value is None:
+            return "processing"
+        elif self._approval_response is None:
+            return "waiting_for_approval"
+        else:
+            return "approved" if self._approval_response.get("approved") else "rejected"
+
+    @workflow.query
+    def get_graph_ascii(self) -> str:
+        """Query to get ASCII art visualization of graph execution progress.
+
+        Returns an ASCII diagram showing which nodes have completed,
+        which is currently executing/interrupted, and which are pending.
+        """
+        if self._app is None:
+            return "Graph not yet initialized"
+        return self._app.get_graph_ascii()
+
+    @workflow.query
+    def get_graph_mermaid(self) -> str:
+        """Query to get Mermaid diagram of graph execution progress.
+
+        Returns a Mermaid flowchart with nodes colored by status:
+        - Green: completed nodes
+        - Yellow: current/interrupted node
+        - Gray: pending nodes
+
+        Can be rendered in GitHub, Notion, or any Mermaid-compatible viewer.
+        """
+        if self._app is None:
+            return "Graph not yet initialized"
```

nit: This should never happen, right?  Why not fail hard?